<a href="https://colab.research.google.com/github/Melvinmcrn/DataScience/blob/master/Final/Part2_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [0]:
import pandas as pd
import numpy as np


In [0]:
columns_name = ['age','workclass','fnlwgt','education','education-num','maritial-status','occupation','relationship','race','sex','capital-gain','capital-loss','hours-per-week','native-country','salary']
data = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/adult/adult.data', names=columns_name)
data.head()

age          workclass  fnlwgt  ... hours-per-week  native-country  salary
0   39          State-gov   77516  ...             40   United-States   <=50K
1   50   Self-emp-not-inc   83311  ...             13   United-States   <=50K
2   38            Private  215646  ...             40   United-States   <=50K
3   53            Private  234721  ...             40   United-States   <=50K
4   28            Private  338409  ...             40            Cuba   <=50K

[5 rows x 15 columns]

In [0]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
age                32561 non-null int64
workclass          32561 non-null object
fnlwgt             32561 non-null int64
education          32561 non-null object
education-num      32561 non-null int64
maritial-status    32561 non-null object
occupation         32561 non-null object
relationship       32561 non-null object
race               32561 non-null object
sex                32561 non-null object
capital-gain       32561 non-null int64
capital-loss       32561 non-null int64
hours-per-week     32561 non-null int64
native-country     32561 non-null object
salary             32561 non-null object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [0]:
data.describe()

age        fnlwgt  ...  capital-loss  hours-per-week
count  32561.000000  3.256100e+04  ...  32561.000000    32561.000000
mean      38.581647  1.897784e+05  ...     87.303830       40.437456
std       13.640433  1.055500e+05  ...    402.960219       12.347429
min       17.000000  1.228500e+04  ...      0.000000        1.000000
25%       28.000000  1.178270e+05  ...      0.000000       40.000000
50%       37.000000  1.783560e+05  ...      0.000000       40.000000
75%       48.000000  2.370510e+05  ...      0.000000       45.000000
max       90.000000  1.484705e+06  ...   4356.000000       99.000000

[8 rows x 6 columns]

# Data Preparation

First, strip every data in the dataframe because every data start with a space.

In [0]:
df_obj = data.select_dtypes(['object'])
data[df_obj.columns] = df_obj.apply(lambda x: x.str.strip())
data['sex'][0]

'Male'

Set '?' value to null so we can deal with NaN easily later.

In [0]:
def CheckQuestionValue():
  for c in data.columns:
    if len(data.loc[data[c] == '?']):
      print(c)

In [0]:
CheckQuestionValue()

workclass
occupation
native-country


/usr/local/lib/python3.6/dist-packages/pandas/core/ops/__init__.py:1115: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


In [0]:
mapping_dict ={'?':np.nan}
data = data.replace(mapping_dict)
CheckQuestionValue()

/usr/local/lib/python3.6/dist-packages/pandas/core/ops/__init__.py:1115: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


Change sex data to (0 for Male) and (1 for Female)

In [0]:
mapping_dict = { 'sex': {'Male':0, 'Female':1}}
cleaned_data = data.replace(mapping_dict)
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
age                32561 non-null int64
workclass          30725 non-null object
fnlwgt             32561 non-null int64
education          32561 non-null object
education-num      32561 non-null int64
maritial-status    32561 non-null object
occupation         30718 non-null object
relationship       32561 non-null object
race               32561 non-null object
sex                32561 non-null int64
capital-gain       32561 non-null int64
capital-loss       32561 non-null int64
hours-per-week     32561 non-null int64
native-country     31978 non-null object
salary             32561 non-null object
dtypes: int64(7), object(8)
memory usage: 3.7+ MB


Change target value to 0 (>50K) and 1 (<=50K)

In [0]:
mapping_dict = { 'salary': {'>50K':0, '<=50K':1}}
cleaned_data = cleaned_data.replace(mapping_dict)
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
age                32561 non-null int64
workclass          30725 non-null object
fnlwgt             32561 non-null int64
education          32561 non-null object
education-num      32561 non-null int64
maritial-status    32561 non-null object
occupation         30718 non-null object
relationship       32561 non-null object
race               32561 non-null object
sex                32561 non-null int64
capital-gain       32561 non-null int64
capital-loss       32561 non-null int64
hours-per-week     32561 non-null int64
native-country     31978 non-null object
salary             32561 non-null int64
dtypes: int64(8), object(7)
memory usage: 3.7+ MB


Deal with cat data and make them numeric

In [0]:
cleaned_data['workclass'].value_counts()

Private             22696
Self-emp-not-inc     2541
Local-gov            2093
State-gov            1298
Self-emp-inc         1116
Federal-gov           960
Without-pay            14
Never-worked            7
Name: workclass, dtype: int64

In [0]:
dummy = pd.get_dummies(cleaned_data['workclass'], drop_first=True)
dummy.head(5)

Local-gov  Never-worked  Private  ...  Self-emp-not-inc  State-gov  Without-pay
0          0             0        0  ...                 0          1            0
1          0             0        0  ...                 1          0            0
2          0             0        1  ...                 0          0            0
3          0             0        1  ...                 0          0            0
4          0             0        1  ...                 0          0            0

[5 rows x 7 columns]

In [0]:
cleaned_data = pd.concat([cleaned_data, dummy], axis=1)
cleaned_data = cleaned_data.drop(['workclass'], axis=1)
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 21 columns):
age                 32561 non-null int64
fnlwgt              32561 non-null int64
education           32561 non-null object
education-num       32561 non-null int64
maritial-status     32561 non-null object
occupation          30718 non-null object
relationship        32561 non-null object
race                32561 non-null object
sex                 32561 non-null int64
capital-gain        32561 non-null int64
capital-loss        32561 non-null int64
hours-per-week      32561 non-null int64
native-country      31978 non-null object
salary              32561 non-null int64
Local-gov           32561 non-null uint8
Never-worked        32561 non-null uint8
Private             32561 non-null uint8
Self-emp-inc        32561 non-null uint8
Self-emp-not-inc    32561 non-null uint8
State-gov           32561 non-null uint8
Without-pay         32561 non-null uint8
dtypes: int64(8), object

In [0]:
cleaned_data['education'].value_counts()

HS-grad         10501
Some-college     7291
Bachelors        5355
Masters          1723
Assoc-voc        1382
11th             1175
Assoc-acdm       1067
10th              933
7th-8th           646
Prof-school       576
9th               514
12th              433
Doctorate         413
5th-6th           333
1st-4th           168
Preschool          51
Name: education, dtype: int64

In [0]:
dummy = pd.get_dummies(cleaned_data['education'], drop_first=True)
cleaned_data = pd.concat([cleaned_data, dummy], axis=1)
cleaned_data = cleaned_data.drop(['education'], axis=1)

In [0]:
cleaned_data['maritial-status'].value_counts()

Married-civ-spouse       14976
Never-married            10683
Divorced                  4443
Separated                 1025
Widowed                    993
Married-spouse-absent      418
Married-AF-spouse           23
Name: maritial-status, dtype: int64

In [0]:
dummy = pd.get_dummies(cleaned_data['maritial-status'], drop_first=True)
cleaned_data = pd.concat([cleaned_data, dummy], axis=1)
cleaned_data = cleaned_data.drop(['maritial-status'], axis=1)

In [0]:
cleaned_data['occupation'].value_counts()

Prof-specialty       4140
Craft-repair         4099
Exec-managerial      4066
Adm-clerical         3770
Sales                3650
Other-service        3295
Machine-op-inspct    2002
Transport-moving     1597
Handlers-cleaners    1370
Farming-fishing       994
Tech-support          928
Protective-serv       649
Priv-house-serv       149
Armed-Forces            9
Name: occupation, dtype: int64

In [0]:
dummy = pd.get_dummies(cleaned_data['occupation'], drop_first=True)
cleaned_data = pd.concat([cleaned_data, dummy], axis=1)
cleaned_data = cleaned_data.drop(['occupation'], axis=1)

In [0]:
cleaned_data['relationship'].value_counts()

Husband           13193
Not-in-family      8305
Own-child          5068
Unmarried          3446
Wife               1568
Other-relative      981
Name: relationship, dtype: int64

In [0]:
dummy = pd.get_dummies(cleaned_data['relationship'], drop_first=True)
cleaned_data = pd.concat([cleaned_data, dummy], axis=1)
cleaned_data = cleaned_data.drop(['relationship'], axis=1)

In [0]:
cleaned_data['race'].value_counts()

White                 27816
Black                  3124
Asian-Pac-Islander     1039
Amer-Indian-Eskimo      311
Other                   271
Name: race, dtype: int64

In [0]:
dummy = pd.get_dummies(cleaned_data['race'], drop_first=True)
cleaned_data = pd.concat([cleaned_data, dummy], axis=1)
cleaned_data = cleaned_data.drop(['race'], axis=1)

In [0]:
cleaned_data['native-country'].value_counts()

United-States                 29170
Mexico                          643
Philippines                     198
Germany                         137
Canada                          121
Puerto-Rico                     114
El-Salvador                     106
India                           100
Cuba                             95
England                          90
Jamaica                          81
South                            80
China                            75
Italy                            73
Dominican-Republic               70
Vietnam                          67
Guatemala                        64
Japan                            62
Poland                           60
Columbia                         59
Taiwan                           51
Haiti                            44
Iran                             43
Portugal                         37
Nicaragua                        34
Peru                             31
Greece                           29
France                      

In [0]:
#drop native-country because there are too many value
cleaned_data = cleaned_data.drop(['native-country'], axis=1)

In [0]:
cleaned_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 58 columns):
age                      32561 non-null int64
fnlwgt                   32561 non-null int64
education-num            32561 non-null int64
sex                      32561 non-null int64
capital-gain             32561 non-null int64
capital-loss             32561 non-null int64
hours-per-week           32561 non-null int64
salary                   32561 non-null int64
Local-gov                32561 non-null uint8
Never-worked             32561 non-null uint8
Private                  32561 non-null uint8
Self-emp-inc             32561 non-null uint8
Self-emp-not-inc         32561 non-null uint8
State-gov                32561 non-null uint8
Without-pay              32561 non-null uint8
11th                     32561 non-null uint8
12th                     32561 non-null uint8
1st-4th                  32561 non-null uint8
5th-6th                  32561 non-null uint8
7th-8th          

In [0]:
cleaned_data.isnull().sum()

age                      0
fnlwgt                   0
education-num            0
sex                      0
capital-gain             0
capital-loss             0
hours-per-week           0
salary                   0
Local-gov                0
Never-worked             0
Private                  0
Self-emp-inc             0
Self-emp-not-inc         0
State-gov                0
Without-pay              0
11th                     0
12th                     0
1st-4th                  0
5th-6th                  0
7th-8th                  0
9th                      0
Assoc-acdm               0
Assoc-voc                0
Bachelors                0
Doctorate                0
HS-grad                  0
Masters                  0
Preschool                0
Prof-school              0
Some-college             0
Married-AF-spouse        0
Married-civ-spouse       0
Married-spouse-absent    0
Never-married            0
Separated                0
Widowed                  0
Armed-Forces             0
C

In [0]:
nonnull_data = cleaned_data.dropna()
nonnull_data.isnull().sum()

age                      0
fnlwgt                   0
education-num            0
sex                      0
capital-gain             0
capital-loss             0
hours-per-week           0
salary                   0
Local-gov                0
Never-worked             0
Private                  0
Self-emp-inc             0
Self-emp-not-inc         0
State-gov                0
Without-pay              0
11th                     0
12th                     0
1st-4th                  0
5th-6th                  0
7th-8th                  0
9th                      0
Assoc-acdm               0
Assoc-voc                0
Bachelors                0
Doctorate                0
HS-grad                  0
Masters                  0
Preschool                0
Prof-school              0
Some-college             0
Married-AF-spouse        0
Married-civ-spouse       0
Married-spouse-absent    0
Never-married            0
Separated                0
Widowed                  0
Armed-Forces             0
C

In [0]:
nonnull_data.shape

(32561, 58)

In [0]:
data = nonnull_data

# Divide data into train/test data

In [0]:
from sklearn.model_selection import train_test_split

In [0]:
X_train, X_test, y_train, y_test = train_test_split(data.drop('salary', axis=1), data['salary'], test_size=0.3, stratify=data['salary'], random_state=1234)

# Choosing best model and parameter

In [0]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report

## RandomForest

In [0]:
parameters = {'min_samples_leaf':[1,2,3,4,5,6,7,8,9,10], 'max_depth':[None, 1,2,3,4,5,6,7,8,9,10], 'criterion':['gini','entropy'], 'min_samples_split':[2,3,4,5,6,7,8,9,10]}

In [0]:
ran_randomForest = RandomizedSearchCV(RandomForestClassifier(random_state=1234), parameters, random_state=1234)
ran_randomForest.fit(X_train, y_train)
ran_randomForest.best_params_

In [0]:
ran_randomForest.best_score_

0.8614863520661915

In [0]:
# grid_randomForest = GridSearchCV(RandomForestClassifier(), parameters, cv=5, scoring='accuracy')
# grid_randomForest.fit(X_train, y_train)
# grid_randomForest.best_params_

KeyboardInterrupt: ignored

In [0]:
# grid_randomForest.best_score_

## kNN

In [0]:
parameters = {'n_neighbors':[3,5,10], 'weights':['uniform','distance'], }

In [0]:
ran_knn = RandomizedSearchCV(KNeighborsClassifier(algorithm='auto'), parameters, random_state=1234)
ran_knn.fit(X_train, y_train)
ran_knn.best_params_

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_search.py:281: UserWarning: The total space of parameters 6 is smaller than n_iter=10. Running 6 iterations. For exhaustive searches, use GridSearchCV.
  % (grid_size, self.n_iter, grid_size), UserWarning)


{'n_neighbors': 10, 'weights': 'uniform'}

In [0]:
ran_knn.best_score_

0.786591787976875

In [0]:
grid_knn = GridSearchCV(KNeighborsClassifier(algorithm='auto'), parameters, cv=5, scoring='accuracy')
grid_knn.fit(X_train, y_train)
grid_knn.best_params_

{'n_neighbors': 10, 'weights': 'uniform'}

In [0]:
grid_knn.best_score_

0.786591787976875

## MLPClassifier

In [0]:
parameters = {'activation':['identity', 'logistic', 'tanh', 'relu'],'solver':['lbfgs', 'sgd', 'adam']}

In [0]:
ran_mlp = RandomizedSearchCV(MLPClassifier(random_state=1234), parameters, random_state=1234)
ran_mlp.fit(X_train, y_train)
ran_mlp.best_params_

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


{'activation': 'identity', 'solver': 'adam'}

In [0]:
ran_mlp.best_score_

0.8105918491898093

In [0]:
grid_mlp = GridSearchCV(MLPClassifier(random_state=1234), parameters, cv=5, scoring='accuracy')
grid_mlp.fit(X_train, y_train)
grid_mlp.best_params_

/usr/local/lib/python3.6/dist-packages/sklearn/utils/extmath.py:151: RuntimeWarning: overflow encountered in matmul
  ret = a @ b
/usr/local/lib/python3.6/dist-packages/sklearn/utils/extmath.py:151: RuntimeWarning: invalid value encountered in matmul
  ret = a @ b
/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/preprocessing/_label.py:768: RuntimeWarning: invalid value encountered in greater
  y = np.array(y > threshold, dtype=np.int)
/usr/local/lib/python3.6/dist-packages/sklearn/utils/extmath.py:151: RuntimeWarning: overflow encountered in matmul
  ret = a @ b
/usr/local/lib/python3.6/dist-packages/sklearn/utils/extmath.py:151: RuntimeWarning: invalid value encountered in matmul
  ret = a @ b
/usr/local/lib/python3.6/dist-packages/sklearn/

{'activation': 'identity', 'solver': 'adam'}

In [0]:
grid_mlp.best_score_

0.8105918491898093

# Train model

As we can see, the best model is Random Forest with parameter as {'criterion': 'entropy',
 'max_depth': None,
 'min_samples_leaf': 10,
 'min_samples_split': 8} and accuracy = 0.8614863520661915

In [0]:
model = RandomForestClassifier(random_state=1234, criterion='entropy', max_depth=None, min_samples_leaf=10, min_samples_split=8)
model.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='entropy', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=10, min_samples_split=8,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=1234,
                       verbose=0, warm_start=False)

In [0]:
y_pred = model.predict(X_test)

In [0]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.78      0.57      0.66      2352
           1       0.88      0.95      0.91      7417

    accuracy                           0.86      9769
   macro avg       0.83      0.76      0.79      9769
weighted avg       0.85      0.86      0.85      9769

